File: NFLOutcomePredictor/NFLOutcomePredictor/ErrorHandling.swift
import SwiftUI
import Combine

/// Global error container for unexpected exceptions.
@MainActor
class ErrorHandler: ObservableObject {
    @Published var currentError: AppError?

    static let shared = ErrorHandler()

    private init() {}

    func handle(_ error: Error, context: String = "") {
        currentError = AppError(
            error: error,
            context: context,
            timestamp: Date()
        )
    }

    func clear() {
        currentError = nil
    }
}

/// Wrapped error with context for display.
struct AppError: Identifiable {
    let id = UUID()
    let error: Error
    let context: String
    let timestamp: Date

    var localizedDescription: String {
        error.localizedDescription
    }

    var fullDescription: String {
        """
        Error: \(error.localizedDescription)
        Context: \(context.isEmpty ? "None" : context)
        Time: \(timestamp.formatted(date: .long, time: .standard))
        Type: \(String(describing: type(of: error)))
        """
    }

    var stackTrace: String {
        // In Swift, we don't have direct access to stack traces like in other languages
        // but we can show the error details
        """
        \(fullDescription)

        Technical Details:
        \(String(reflecting: error))
        """
    }
}

/// Error view overlay showing "Awe Snap this is Bull Shark".
struct ErrorOverlay: View {
    let error: AppError
    let onDismiss: () -> Void

    @State private var showDetails = false

    var body: some View {
        ZStack {
            // Dimmed background
            Color.black.opacity(0.4)
                .ignoresSafeArea()

            VStack(spacing: 0) {
                if showDetails {
                    // Details view
                    ErrorDetailsView(error: error, onClose: {
                        showDetails = false
                    })
                } else {
                    // Simple error message
                    ErrorMessageView(
                        error: error,
                        onClose: onDismiss,
                        onShowDetails: {
                            showDetails = true
                        }
                    )
                }
            }
        }
    }
}

/// Simple error message with Close and Show Details buttons.
struct ErrorMessageView: View {
    let error: AppError
    let onClose: () -> Void
    let onShowDetails: () -> Void

    @State private var showFirstMessage = false
    @State private var showShark = false
    @State private var showSecondMessage = false

    var body: some View {
        VStack(spacing: 24) {
            // Animated messages and shark
            VStack(spacing: 12) {
                // First message: "Awe Snap, Something bad happened!"
                if showFirstMessage {
                    Text("Awe Snap, Something bad happened!")
                        .font(.title2)
                        .fontWeight(.bold)
                        .multilineTextAlignment(.center)
                        .transition(.asymmetric(
                            insertion: .scale.combined(with: .opacity),
                            removal: .move(edge: .leading).combined(with: .opacity)
                        ))
                }

                // Shark with nose ring
                if showShark {
                    SharkWithNoseRing(size: 80)
                        .transition(.scale.combined(with: .opacity))
                }

                // Second message: "This is Bull Shark!"
                if showSecondMessage {
                    Text("This is Bull Shark!")
                        .font(.title2)
                        .fontWeight(.bold)
                        .multilineTextAlignment(.center)
                        .transition(.scale.combined(with: .opacity))
                }

                // Error description (always visible after animations)
                if showSecondMessage {
                    Text(error.localizedDescription)
                        .font(.body)
                        .foregroundColor(.secondary)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal)
                        .transition(.opacity)
                }
            }
            .frame(minHeight: 200)

            // Buttons (always visible)
            VStack(spacing: 12) {
                Button(action: onClose) {
                    Text("Close")
                        .font(.headline)
                        .frame(maxWidth: .infinity)
                        .padding()
                        .background(Color.accentColor)
                        .foregroundColor(.white)
                        .cornerRadius(12)
                }

                Button(action: onShowDetails) {
                    HStack {
                        Image(systemName: "info.circle")
                        Text("Show Details")
                    }
                    .font(.subheadline)
                    .frame(maxWidth: .infinity)
                    .padding()
                    .background(Color(.systemGray5))
                    .foregroundColor(.primary)
                    .cornerRadius(12)
                }
            }
            .padding(.horizontal, 24)
        }
        .padding(32)
        .background(
            RoundedRectangle(cornerRadius: 20)
                .fill(Color(.systemBackground))
                .shadow(color: .black.opacity(0.2), radius: 20, x: 0, y: 10)
        )
        .padding(40)
        .onAppear {
            startAnimation()
        }
    }

    private func startAnimation() {
        // Show first message immediately
        withAnimation(.easeOut(duration: 0.4)) {
            showFirstMessage = true
        }

        // Hide first message and show shark after 1.2 seconds
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.2) {
            withAnimation(.easeIn(duration: 0.3)) {
                showFirstMessage = false
            }

            DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
                withAnimation(.spring(response: 0.6, dampingFraction: 0.7)) {
                    showShark = true
                }
            }
        }

        // Show second message after 2.0 seconds
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            withAnimation(.easeOut(duration: 0.4)) {
                showSecondMessage = true
            }
        }
    }
}

/// Detailed error view with full stack trace.
struct ErrorDetailsView: View {
    let error: AppError
    let onClose: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            // Header
            HStack {
                VStack(alignment: .leading, spacing: 4) {
                    Text("Error Details")
                        .font(.title2)
                        .fontWeight(.bold)

                    Text(error.timestamp.formatted(date: .abbreviated, time: .standard))
                        .font(.caption)
                        .foregroundColor(.secondary)
                }

                Spacer()

                Button(action: onClose) {
                    Image(systemName: "xmark.circle.fill")
                        .font(.title2)
                        .foregroundColor(.secondary)
                }
            }
            .padding()
            .background(Color(.systemGray6))

            // Scrollable details
            ScrollView {
                VStack(alignment: .leading, spacing: 16) {
                    // Error message
                    DetailSection(title: "Error Message") {
                        Text(error.localizedDescription)
                            .font(.body)
                    }

                    // Context
                    if !error.context.isEmpty {
                        DetailSection(title: "Context") {
                            Text(error.context)
                                .font(.body)
                        }
                    }

                    // Stack trace
                    DetailSection(title: "Stack Trace") {
                        Text(error.stackTrace)
                            .font(.system(.caption, design: .monospaced))
                            .textSelection(.enabled)
                    }

                    // Copy button
                    Button(action: {
                        UIPasteboard.general.string = error.stackTrace
                    }) {
                        HStack {
                            Image(systemName: "doc.on.doc")
                            Text("Copy Error Details")
                        }
                        .font(.subheadline)
                        .frame(maxWidth: .infinity)
                        .padding()
                        .background(Color.accentColor)
                        .foregroundColor(.white)
                        .cornerRadius(12)
                    }
                }
                .padding()
            }
        }
        .frame(maxWidth: .infinity, maxHeight: 600)
        .background(
            RoundedRectangle(cornerRadius: 20)
                .fill(Color(.systemBackground))
                .shadow(color: .black.opacity(0.2), radius: 20, x: 0, y: 10)
        )
        .padding(40)
    }
}

/// Reusable detail section.
struct DetailSection<Content: View>: View {
    let title: String
    @ViewBuilder let content: () -> Content

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(title)
                .font(.headline)
                .foregroundColor(.primary)

            content()
                .padding()
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(Color(.systemGray6))
                .cornerRadius(8)
        }
    }
}

/// View modifier to handle errors globally.
struct ErrorHandlingModifier: ViewModifier {
    @ObservedObject var errorHandler = ErrorHandler.shared

    func body(content: Content) -> some View {
        content
            .overlay {
                if let error = errorHandler.currentError {
                    ErrorOverlay(
                        error: error,
                        onDismiss: {
                            errorHandler.clear()
                        }
                    )
                    .transition(.opacity)
                    .zIndex(999)
                }
            }
    }
}

extension View {
    /// Adds global error handling to a view.
    func withErrorHandling() -> some View {
        modifier(ErrorHandlingModifier())
    }
}

/// Helper for wrapping async operations with error handling.
extension View {
    func handleErrors(context: String = "", perform: @escaping () async throws -> Void) -> some View {
        Task {
            do {
                try await perform()
            } catch {
                await MainActor.run {
                    ErrorHandler.shared.handle(error, context: context)
                }
            }
        }
        return self
    }
}

File: StatSharkAndroid/app/src/main/kotlin/com/statshark/nfl/ui/components/ErrorHandling.kt
package com.statshark.nfl.ui.components

import androidx.compose.animation.*
import androidx.compose.animation.core.*
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Info
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import kotlinx.coroutines.delay
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * App Error data class
 * Wrapped error with context for display
 */
data class AppError(
    val error: Throwable,
    val context: String = "",
    val timestamp: Date = Date()
) {
    val localizedDescription: String
        get() = error.message ?: "Unknown error occurred"

    val fullDescription: String
        get() {
            val dateFormat = SimpleDateFormat("MMMM dd, yyyy 'at' hh:mm:ss a", Locale.US)
            return """
                Error: $localizedDescription
                Context: ${context.ifEmpty { "None" }}
                Time: ${dateFormat.format(timestamp)}
                Type: ${error::class.simpleName}
            """.trimIndent()
        }

    val stackTrace: String
        get() = """
            $fullDescription

            Technical Details:
            ${error.stackTraceToString()}
        """.trimIndent()
}

/**
 * Shark with Nose Ring Component
 * Matching iOS SharkWithNoseRing
 */
@Composable
fun SharkWithNoseRing(
    size: Float = 80f,
    modifier: Modifier = Modifier
) {
    val infiniteTransition = rememberInfiniteTransition(label = "shark_animation")
    val rotation by infiniteTransition.animateFloat(
        initialValue = -5f,
        targetValue = 5f,
        animationSpec = infiniteRepeatable(
            animation = tween(2000, easing = LinearEasing),
            repeatMode = RepeatMode.Reverse
        ),
        label = "rotation"
    )

    Canvas(
        modifier = modifier.size(size.dp)
    ) {
        val centerX = size / 2
        val centerY = size / 2

        // Draw shark body (simple shark silhouette)
        val sharkPath = Path().apply {
            // Body
            moveTo(centerX * 0.3f, centerY)
            cubicTo(
                centerX * 0.3f, centerY * 0.5f,
                centerX * 0.7f, centerY * 0.5f,
                centerX * 1.5f, centerY
            )
            cubicTo(
                centerX * 0.7f, centerY * 1.5f,
                centerX * 0.3f, centerY * 1.5f,
                centerX * 0.3f, centerY
            )
            // Tail fin
            moveTo(centerX * 0.2f, centerY)
            lineTo(centerX * 0.1f, centerY * 0.5f)
            lineTo(centerX * 0.3f, centerY)
            // Dorsal fin
            moveTo(centerX * 0.8f, centerY * 0.6f)
            lineTo(centerX, centerY * 0.2f)
            lineTo(centerX * 1.2f, centerY * 0.6f)
        }

        // Draw shark
        drawPath(
            path = sharkPath,
            color = Color(0xFF607D8B), // Blue-gray shark color
            style = Stroke(width = 3f)
        )

        // Draw nose ring
        val noseX = centerX * 1.5f
        val noseY = centerY * 1.1f
        drawCircle(
            color = Color(0xFFFFD700), // Gold color for ring
            radius = 4f,
            center = Offset(noseX, noseY),
            style = Stroke(width = 2f)
        )
    }
}

/**
 * Error Overlay
 * Shows "Awe Snap this is Bull Shark" error message with animation
 * Matching iOS ErrorOverlay
 */
@Composable
fun ErrorOverlay(
    error: AppError,
    onDismiss: () -> Unit
) {
    var showDetails by remember { mutableStateOf(false) }

    Dialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(
            dismissOnBackPress = true,
            dismissOnClickOutside = true
        )
    ) {
        Box(
            modifier = Modifier.fillMaxSize(),
            contentAlignment = Alignment.Center
        ) {
            if (showDetails) {
                ErrorDetailsView(
                    error = error,
                    onClose = { showDetails = false }
                )
            } else {
                ErrorMessageView(
                    error = error,
                    onClose = onDismiss,
                    onShowDetails = { showDetails = true }
                )
            }
        }
    }
}

/**
 * Simple error message with Close and Show Details buttons
 * Matching iOS ErrorMessageView
 */
@Composable
private fun ErrorMessageView(
    error: AppError,
    onClose: () -> Unit,
    onShowDetails: () -> Unit
) {
    var showFirstMessage by remember { mutableStateOf(false) }
    var showShark by remember { mutableStateOf(false) }
    var showSecondMessage by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        // Show first message immediately
        showFirstMessage = true
        delay(1200)

        // Hide first message and show shark
        showFirstMessage = false
        delay(300)
        showShark = true
        delay(800)

        // Show second message
        showSecondMessage = true
    }

    Card(
        modifier = Modifier
            .fillMaxWidth(0.9f)
            .padding(40.dp)
            .shadow(20.dp, RoundedCornerShape(20.dp)),
        shape = RoundedCornerShape(20.dp),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surface
        )
    ) {
        Column(
            modifier = Modifier
                .padding(32.dp)
                .fillMaxWidth(),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(24.dp)
        ) {
            // Animated messages and shark
            Box(
                modifier = Modifier.height(200.dp),
                contentAlignment = Alignment.Center
            ) {
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(12.dp)
                ) {
                    // First message: "Awe Snap, Something bad happened!"
                    AnimatedVisibility(
                        visible = showFirstMessage,
                        enter = scaleIn() + fadeIn(),
                        exit = slideOutHorizontally() + fadeOut()
                    ) {
                        Text(
                            text = "Awe Snap, Something bad happened!",
                            style = MaterialTheme.typography.headlineSmall,
                            fontWeight = FontWeight.Bold,
                            textAlign = TextAlign.Center
                        )
                    }

                    // Shark with nose ring
                    AnimatedVisibility(
                        visible = showShark,
                        enter = scaleIn(
                            animationSpec = spring(
                                dampingRatio = 0.7f,
                                stiffness = Spring.StiffnessMedium
                            )
                        ) + fadeIn()
                    ) {
                        SharkWithNoseRing(size = 80f)
                    }

                    // Second message: "This is Bull Shark!"
                    AnimatedVisibility(
                        visible = showSecondMessage,
                        enter = scaleIn() + fadeIn()
                    ) {
                        Column(
                            horizontalAlignment = Alignment.CenterHorizontally,
                            verticalArrangement = Arrangement.spacedBy(8.dp)
                        ) {
                            Text(
                                text = "This is Bull Shark!",
                                style = MaterialTheme.typography.headlineSmall,
                                fontWeight = FontWeight.Bold,
                                textAlign = TextAlign.Center
                            )
                            Text(
                                text = error.localizedDescription,
                                style = MaterialTheme.typography.bodyMedium,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                                textAlign = TextAlign.Center,
                                modifier = Modifier.padding(horizontal = 16.dp)
                            )
                        }
                    }
                }
            }

            // Buttons
            Column(
                modifier = Modifier.fillMaxWidth(),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                Button(
                    onClick = onClose,
                    modifier = Modifier.fillMaxWidth(),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = MaterialTheme.colorScheme.primary
                    )
                ) {
                    Text(
                        text = "Close",
                        style = MaterialTheme.typography.titleMedium
                    )
                }

                OutlinedButton(
                    onClick = onShowDetails,
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Icon(
                        imageVector = Icons.Filled.Info,
                        contentDescription = null,
                        modifier = Modifier.size(18.dp)
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Show Details")
                }
            }
        }
    }
}

/**
 * Detailed error view with full stack trace
 * Matching iOS ErrorDetailsView
 */
@Composable
private fun ErrorDetailsView(
    error: AppError,
    onClose: () -> Unit
) {
    val clipboardManager = LocalClipboardManager.current
    val dateFormat = remember { SimpleDateFormat("MMM dd, yyyy hh:mm a", Locale.US) }

    Card(
        modifier = Modifier
            .fillMaxWidth(0.9f)
            .fillMaxHeight(0.8f)
            .padding(40.dp)
            .shadow(20.dp, RoundedCornerShape(20.dp)),
        shape = RoundedCornerShape(20.dp),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surface
        )
    ) {
        Column(
            modifier = Modifier.fillMaxSize()
        ) {
            // Header
            Surface(
                color = MaterialTheme.colorScheme.surfaceVariant
            ) {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(16.dp),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Column {
                        Text(
                            text = "Error Details",
                            style = MaterialTheme.typography.headlineSmall,
                            fontWeight = FontWeight.Bold
                        )
                        Text(
                            text = dateFormat.format(error.timestamp),
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }

                    IconButton(onClick = onClose) {
                        Icon(
                            imageVector = Icons.Filled.Close,
                            contentDescription = "Close"
                        )
                    }
                }
            }

            // Scrollable details
            Column(
                modifier = Modifier
                    .weight(1f)
                    .verticalScroll(rememberScrollState())
                    .padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                // Error message
                DetailSection(title = "Error Message") {
                    Text(
                        text = error.localizedDescription,
                        style = MaterialTheme.typography.bodyMedium
                    )
                }

                // Context
                if (error.context.isNotEmpty()) {
                    DetailSection(title = "Context") {
                        Text(
                            text = error.context,
                            style = MaterialTheme.typography.bodyMedium
                        )
                    }
                }

                // Stack trace
                DetailSection(title = "Stack Trace") {
                    Text(
                        text = error.stackTrace,
                        style = MaterialTheme.typography.bodySmall.copy(
                            fontFamily = androidx.compose.ui.text.font.FontFamily.Monospace
                        )
                    )
                }

                // Copy button
                Button(
                    onClick = {
                        clipboardManager.setText(AnnotatedString(error.stackTrace))
                    },
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Icon(
                        painter = androidx.compose.ui.res.painterResource(android.R.drawable.ic_menu_upload),
                        contentDescription = null,
                        modifier = Modifier.size(18.dp)
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Copy Error Details")
                }
            }
        }
    }
}

/**
 * Reusable detail section
 * Matching iOS DetailSection
 */
@Composable
private fun DetailSection(
    title: String,
    content: @Composable () -> Unit
) {
    Column(
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = title,
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.SemiBold
        )
        Surface(
            modifier = Modifier.fillMaxWidth(),
            shape = RoundedCornerShape(8.dp),
            color = MaterialTheme.colorScheme.surfaceVariant
        ) {
            Box(modifier = Modifier.padding(12.dp)) {
                content()
            }
        }
    }
}

/**
 * Error Handler Singleton
 * Global error container for unexpected exceptions
 * Matching iOS ErrorHandler
 */
object ErrorHandler {
    private val _currentError = mutableStateOf<AppError?>(null)
    val currentError: AppError? by _currentError

    fun handle(error: Throwable, context: String = "") {
        _currentError.value = AppError(
            error = error,
            context = context,
            timestamp = Date()
        )
    }

    fun clear() {
        _currentError.value = null
    }
}

/**
 * Composable function to show error overlay when error exists
 * Use this at the root of your app to handle errors globally
 * Matching iOS .withErrorHandling() modifier
 */
@Composable
fun ErrorHandlingOverlay() {
    val error = ErrorHandler.currentError

    if (error != null) {
        ErrorOverlay(
            error = error,
            onDismiss = { ErrorHandler.clear() }
        )
    }
}
